node/src/node/subscriptions.rs

```diff
@@ -0,0 +1,277 @@
+                .insert_relaxed(height..=height)
+                .expect("single item range should be valid");
+        }
+    }
+
+    pub(crate) fn subscribe(&self) -> broadcast::Receiver<u64> {
+        self.sender.subscribe()
+    }
+
+    pub(crate) fn signal_inserted(&mut self, range: BlockRange) {
+        let Some(last_sent_height) = self.last_sent_height else {
+            // header insertions before head height is initialised
+            // this shouldn't happen, ignore it
+            return;
+        };
+
+        if &last_sent_height > range.end() {
+            // we don't care about past ranges syncer is fetching
+            return;
+        }
+
+        self.pending
+            .insert_relaxed(&range)
+            .expect("inserted block range should be valid");
+
+        let Some(first_pending) = self.pending.tail() else {
+            return;
+        };
+
+        if last_sent_height + 1 == first_pending {
+            let to_send = self
+                .pending
+                .pop_tail_range()
+                .expect("pending shouldn't be empty");
+            self.last_sent_height = Some(*to_send.end());
+            for h in to_send {
+                let _ = self.sender.send(h);
+            }
+        }
+    }
+}
+
+pub(crate) async fn forward_new_headers<S: Store>(
+    syncer: Arc<Syncer<S>>,
+    store: Arc<S>,
+    tx: mpsc::Sender<Result<ExtendedHeader, SubscriptionError>>,
+) {
+    let mut heights_receiver = syncer.subscribe_heights().await.unwrap();
+
+    loop {
+        let height = match heights_receiver.recv().await {
+            Ok(height) => height,
+            Err(RecvError::Lagged(skipped)) => {
+                let _ = tx.send(Err(SubscriptionError::Lagged(skipped))).await;
+                continue;
+            }
+            Err(RecvError::Closed) => {
+                error!("unexpected heigth stream close");
+                return;
+            }
+        };
+
+        let header_or_error = match store.get_by_height(height).await {
+            Ok(h) => Ok(h),
+            Err(e) => Err(SubscriptionError::Node {
+                height,
+                source: NodeError::Store(e),
+            }),
+        };
+
+        if tx.send(header_or_error).await.is_err() {
+            return; // receiver dropped
+        }
+    }
+}
+
+pub(crate) async fn forward_new_blobs<S: Store>(
+    namespace: Namespace,
+    tx: mpsc::Sender<Result<BlobsAtHeight, SubscriptionError>>,
+    syncer: Arc<Syncer<S>>,
+    store: Arc<S>,
+    p2p: Arc<P2p>,
+) {
+    let mut heights_receiver = syncer.subscribe_heights().await.unwrap();
+
+    loop {
+        let height = match heights_receiver.recv().await {
+            Ok(height) => height,
+            Err(RecvError::Lagged(skipped)) => {
+                let _ = tx.send(Err(SubscriptionError::Lagged(skipped))).await;
+                continue;
+            }
+            Err(RecvError::Closed) => {
+                error!("unexpected heigth stream close");
```

Comment:
Not really an unexpected error when `Syncer` is manually stopped. I prefer remove this log to be removed. Do it in other functions too